File: python/src/pi_mono/agent/harness/session/jsonl_storage.py
"""JSONL session storage implementation."""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass
from datetime import datetime
from typing import Any

from pi_mono.agent.harness.types import (
    JsonlSessionMetadata,
    SessionError,
    SessionTreeEntry,
    MessageEntry,
    ThinkingLevelChangeEntry,
    ModelChangeEntry,
    ActiveToolsChangeEntry,
    CompactionEntry,
    BranchSummaryEntry,
    CustomEntry,
    CustomMessageEntry,
    LabelEntry,
    SessionInfoEntry,
    LeafEntry,
)


@dataclass
class SessionHeader:
    type: str = "session"
    version: int = 3
    id: str = ""
    timestamp: str = ""
    cwd: str = ""
    parent_session: str | None = None


def _get_type(entry: Any) -> str:
    if isinstance(entry, dict):
        return entry.get("type", "")
    return getattr(entry, "type", "")


def _get_id(entry: Any) -> str:
    if isinstance(entry, dict):
        return entry.get("id", "")
    return getattr(entry, "id", "")


def _get_parent_id(entry: Any) -> str | None:
    if isinstance(entry, dict):
        return entry.get("parentId", entry.get("parent_id"))
    return getattr(entry, "parent_id", getattr(entry, "parentId", None))


def _get_target_id(entry: Any) -> str | None:
    if isinstance(entry, dict):
        return entry.get("targetId", entry.get("target_id"))
    return getattr(entry, "target_id", getattr(entry, "targetId", None))


def _get_label(entry: Any) -> str | None:
    if isinstance(entry, dict):
        return entry.get("label")
    return getattr(entry, "label", None)


def _serialize_entry(entry: Any) -> dict:
    if isinstance(entry, dict):
        return entry

    res = {}
    res["type"] = entry.type
    res["id"] = entry.id
    res["parentId"] = entry.parent_id
    res["timestamp"] = entry.timestamp

    if entry.type == "message":
        res["message"] = entry.message
    elif entry.type == "thinking_level_change":
        res["thinkingLevel"] = entry.thinking_level
    elif entry.type == "model_change":
        res["provider"] = entry.provider
        res["modelId"] = entry.model_id
    elif entry.type == "active_tools_change":
        res["activeToolNames"] = entry.active_tool_names
    elif entry.type == "compaction":
        res["summary"] = entry.summary
        res["firstKeptEntryId"] = entry.first_kept_entry_id
        res["tokensBefore"] = entry.tokens_before
        res["details"] = entry.details
        res["fromHook"] = entry.from_hook
    elif entry.type == "branch_summary":
        res["fromId"] = entry.from_id
        res["summary"] = entry.summary
        res["details"] = entry.details
        res["fromHook"] = entry.from_hook
    elif entry.type == "custom":
        res["customType"] = entry.custom_type
        res["data"] = entry.data
    elif entry.type == "custom_message":
        res["customType"] = entry.custom_type
        res["content"] = entry.content
        res["details"] = entry.details
        res["display"] = entry.display
    elif entry.type == "label":
        res["targetId"] = entry.target_id
        res["label"] = entry.label
    elif entry.type == "session_info":
        res["name"] = entry.name
    elif entry.type == "leaf":
        res["targetId"] = entry.target_id

    return res


def _deserialize_entry(data: dict) -> Any:
    t = data.get("type")
    base_args = {
        "id": data.get("id", ""),
        "parent_id": data.get("parentId", data.get("parent_id")),
        "timestamp": data.get("timestamp", ""),
    }
    if t == "message":
        return MessageEntry(**base_args, message=data.get("message"))
    elif t == "thinking_level_change":
        return ThinkingLevelChangeEntry(
            **base_args, thinking_level=data.get("thinkingLevel", data.get("thinking_level", ""))
        )
    elif t == "model_change":
        return ModelChangeEntry(
            **base_args,
            provider=data.get("provider", ""),
            model_id=data.get("modelId", data.get("model_id", "")),
        )
    elif t == "active_tools_change":
        return ActiveToolsChangeEntry(
            **base_args,
            active_tool_names=data.get("activeToolNames", data.get("active_tool_names")),
        )
    elif t == "compaction":
        return CompactionEntry(
            **base_args,
            summary=data.get("summary", ""),
            first_kept_entry_id=data.get("firstKeptEntryId", data.get("first_kept_entry_id", "")),
            tokens_before=data.get("tokensBefore", data.get("tokens_before", 0)),
            details=data.get("details"),
            from_hook=data.get("fromHook", data.get("from_hook", False)),
        )
    elif t == "branch_summary":
        return BranchSummaryEntry(
            **base_args,
            from_id=data.get("fromId", data.get("from_id", "")),
            summary=data.get("summary", ""),
            details=data.get("details"),
            from_hook=data.get("fromHook", data.get("from_hook", False)),
        )
    elif t == "custom":
        return CustomEntry(
            **base_args,
            custom_type=data.get("customType", data.get("custom_type", "")),
            data=data.get("data"),
        )
    elif t == "custom_message":
        return CustomMessageEntry(
            **base_args,
            custom_type=data.get("customType", data.get("custom_type", "")),
            content=data.get("content", ""),
            details=data.get("details"),
            display=data.get("display", False),
        )
    elif t == "label":
        return LabelEntry(
            **base_args,
            target_id=data.get("targetId", data.get("target_id", "")),
            label=data.get("label"),
        )
    elif t == "session_info":
        return SessionInfoEntry(**base_args, name=data.get("name"))
    elif t == "leaf":
        return LeafEntry(**base_args, target_id=data.get("targetId", data.get("target_id")))
    else:
        return data


def _invalid_session(file_path: str, message: str, cause: Exception | None = None) -> SessionError:
    return SessionError(
        "invalid_session", f"Invalid JSONL session file {file_path}: {message}", cause
    )


def _invalid_entry(
    file_path: str, line_number: int, message: str, cause: Exception | None = None
) -> SessionError:
    return SessionError(
        "invalid_entry",
        f"Invalid JSONL session file {file_path}: line {line_number} {message}",
        cause,
    )


def parse_header_line(line: str, file_path: str) -> dict:
    try:
        parsed = json.loads(line)
    except json.JSONDecodeError as e:
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: first line is not a valid session header",
            e,
        )

    if not isinstance(parsed, dict):
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: first line is not a valid session header",
        )

    if parsed.get("type") != "session":
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: first line is not a valid session header",
        )
    if parsed.get("version") != 3:
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: unsupported session version",
        )
    if not isinstance(parsed.get("id"), str) or not parsed.get("id"):
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: session header is missing id",
        )
    if not isinstance(parsed.get("timestamp"), str) or not parsed.get("timestamp"):
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: session header is missing timestamp",
        )
    if not isinstance(parsed.get("cwd"), str) or not parsed.get("cwd"):
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: session header is missing cwd",
        )
    if parsed.get("parentSession") is not None and not isinstance(parsed.get("parentSession"), str):
        raise SessionError(
            "invalid_session",
            f"Invalid JSONL session file {file_path}: session header parentSession must be a string",
        )

    return {
        "type": "session",
        "version": 3,
        "id": parsed["id"],
        "timestamp": parsed["timestamp"],
        "cwd": parsed["cwd"],
        "parentSession": parsed.get("parentSession"),
    }


def parse_entry_line(line: str, file_path: str, line_number: int) -> dict:
    try:
        parsed = json.loads(line)
    except json.JSONDecodeError as e:
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} is not valid JSON",
            e,
        )

    if not isinstance(parsed, dict):
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} is not a valid session entry",
        )

    if not isinstance(parsed.get("type"), str):
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} is missing entry type",
        )
    if not isinstance(parsed.get("id"), str) or not parsed.get("id"):
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} is missing entry id",
        )
    if parsed.get("parentId") is not None and not isinstance(parsed.get("parentId"), str):
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} has invalid parentId",
        )
    if not isinstance(parsed.get("timestamp"), str) or not parsed.get("timestamp"):
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} is missing timestamp",
        )
    if (
        parsed.get("type") == "leaf"
        and parsed.get("targetId") is not None
        and not isinstance(parsed.get("targetId"), str)
    ):
        raise SessionError(
            "invalid_entry",
            f"Invalid JSONL session file {file_path}: line {line_number} has invalid targetId",
        )

    return parsed


def _leaf_id_after_entry(entry: Any) -> str | None:
    if _get_type(entry) == "leaf":
        return _get_target_id(entry)
    return _get_id(entry)


def _header_to_session_metadata(header: dict, file_path: str) -> JsonlSessionMetadata:
    return JsonlSessionMetadata(
        id=header["id"],
        created_at=header["timestamp"],
        cwd=header["cwd"],
        path=file_path,
        parent_session_path=header.get("parentSession"),
    )


async def _load_jsonl_session_metadata(
    fs: Any,
    file_path: str,
) -> JsonlSessionMetadata:
    """Load just the metadata from a JSONL session file."""
    lines_result = await fs.read_text_lines(file_path, {"maxLines": 1})
    if not lines_result.ok:
        raise SessionError(
            "storage", f"Failed to read session header {file_path}", lines_result.error
        )
    lines = lines_result.value
    if not lines or not lines[0].strip():
        raise SessionError(
            "invalid_session", f"Invalid JSONL session file {file_path}: missing session header"
        )
    header = parse_header_line(lines[0], file_path)
    return _header_to_session_metadata(header, file_path)


class JsonlSessionStorage:
    """JSONL session storage implementation."""

    def __init__(
        self,
        fs: Any,
        file_path: str,
        header: dict,
        entries: list[SessionTreeEntry],
        leaf_id: str | None,
    ):
        self.fs = fs
        self.file_path = file_path
        self.metadata = _header_to_session_metadata(header, file_path)
        self.entries = entries
        self.by_id = {_get_id(entry): entry for entry in entries}
        self.labels_by_id: dict[str, str] = {}
        self._build_labels_by_id()
        self.current_leaf_id = leaf_id

    def _build_labels_by_id(self) -> None:
        self.labels_by_id = {}
        for entry in self.entries:
            if _get_type(entry) == "label":
                label = _get_label(entry)
                target_id = _get_target_id(entry)
                if target_id is not None:
                    if label:
                        self.labels_by_id[target_id] = label.strip()
                    else:
                        self.labels_by_id.pop(target_id, None)

    @classmethod
    async def open(cls, fs: Any, file_path: str) -> "JsonlSessionStorage":
        """Open an existing JSONL session file."""
        content_result = await fs.read_text_file(file_path)
        if not content_result.ok:
            raise SessionError(
                "not_found", f"Failed to read session {file_path}", content_result.error
            )

        content = content_result.value
        lines = [line for line in content.split("\n") if line.strip()]
        if not lines:
            raise SessionError(
                "invalid_session", f"Invalid JSONL session file {file_path}: missing session header"
            )

        header = parse_header_line(lines[0], file_path)
        entries = []
        leaf_id = None

        for i in range(1, len(lines)):
            entry_dict = parse_entry_line(lines[i], file_path, i + 1)
            entry = _deserialize_entry(entry_dict)
            entries.append(entry)
            leaf_id = _leaf_id_after_entry(entry)

        return cls(fs, file_path, header, entries, leaf_id)

    @classmethod
    async def create(
        cls,
        fs: Any,
        file_path: str,
        options: dict,
    ) -> "JsonlSessionStorage":
        """Create a new JSONL session file."""
        header = {
            "type": "session",
            "version": 3,
            "id": options["sessionId"],
            "timestamp": datetime.now().isoformat(),
            "cwd": options["cwd"],
            "parentSession": options.get("parentSessionPath"),
        }

        write_res = await fs.write_file(file_path, json.dumps(header) + "\n")
        if not write_res.ok:
            raise SessionError("storage", f"Failed to write header to {file_path}", write_res.error)
        return cls(fs, file_path, header, [], None)

    async def get_metadata(self) -> JsonlSessionMetadata:
        return self.metadata

    async def get_leaf_id(self) -> str | None:
        if self.current_leaf_id and self.current_leaf_id not in self.by_id:
            raise SessionError("invalid_session", f"Entry {self.current_leaf_id} not found")
        return self.current_leaf_id

    async def set_leaf_id(self, leaf_id: str | None) -> None:
        if leaf_id is not None and leaf_id not in self.by_id:
            raise SessionError("not_found", f"Entry {leaf_id} not found")

        entry = LeafEntry(
            type="leaf",
            id=generate_entry_id(self.by_id),
            parent_id=self.current_leaf_id,
            timestamp=datetime.now().isoformat(),
            target_id=leaf_id,
        )

        serialized = _serialize_entry(entry)
        json_line = json.dumps(serialized) + "\n"
        append_res = await self.fs.append_file(self.file_path, json_line)
        if not append_res.ok:
            raise SessionError(
                "storage", f"Failed to set leaf id in {self.file_path}", append_res.error
            )

        self.entries.append(entry)
        self.by_id[entry.id] = entry
        self.current_leaf_id = leaf_id

    async def create_entry_id(self) -> str:
        return generate_entry_id(self.by_id)

    async def append_entry(self, entry: SessionTreeEntry) -> None:
        serialized = _serialize_entry(entry)
        json_line = json.dumps(serialized) + "\n"
        append_res = await self.fs.append_file(self.file_path, json_line)
        if not append_res.ok:
            raise SessionError(
                "storage", f"Failed to append entry to {self.file_path}", append_res.error
            )

        deserialized = entry if not isinstance(entry, dict) else _deserialize_entry(entry)
        self.entries.append(deserialized)
        self.by_id[deserialized.id] = deserialized
        self._update_label_cache(deserialized)
        self.current_leaf_id = _leaf_id_after_entry(deserialized)

    async def get_entry(self, id: str) -> SessionTreeEntry | None:
        return self.by_id.get(id)

    async def find_entries(self, type_: str) -> list[SessionTreeEntry]:
        return [entry for entry in self.entries if _get_type(entry) == type_]

    async def get_label(self, id: str) -> str | None:
        return self.labels_by_id.get(id)

    async def get_path_to_root(self, leaf_id: str | None) -> list[SessionTreeEntry]:
        if leaf_id is None:
            return []
        path = []
        current = self.by_id.get(leaf_id)
        if not current:
            raise SessionError("not_found", f"Entry {leaf_id} not found")

        while current:
            path.insert(0, current)
            parent_id = _get_parent_id(current)
            if not parent_id:
                break
            parent = self.by_id.get(parent_id)
            if not parent:
                raise SessionError("invalid_session", f"Entry {parent_id} not found")
            current = parent
        return path

    async def get_entries(self) -> list[SessionTreeEntry]:
        return list(self.entries)

    def _update_label_cache(self, entry: SessionTreeEntry) -> None:
        if _get_type(entry) != "label":
            return
        label = _get_label(entry)
        target_id = _get_target_id(entry)
        if target_id is not None:
            if label:
                self.labels_by_id[target_id] = label.strip()
            else:
                self.labels_by_id.pop(target_id, None)


def generate_entry_id(by_id: dict) -> str:
    """Generate a unique 8-character entry ID."""
    for _ in range(100):
        id_ = str(uuid.uuid4())[:8]
        if id_ not in by_id:
            return id_
    return str(uuid.uuid4())


def leaf_id_after_entry(entry: Any) -> str | None:
    return _leaf_id_after_entry(entry)


loadJsonlSessionMetadata = _load_jsonl_session_metadata

File: packages/agent/src/harness/session/jsonl-storage.ts
import { uuidv7 } from "@earendil-works/pi-ai";
import type {
	FileSystem,
	JsonlSessionMetadata,
	LeafEntry,
	SessionEntryCursorOptions,
	SessionStorage,
	SessionTreeEntry,
} from "../types.ts";
import { SessionError, toError } from "../types.ts";
import { getFileSystemResultOrThrow } from "./repo-utils.ts";

type JsonlSessionStorageFileSystem = Pick<FileSystem, "readTextFile" | "readTextLines" | "writeFile" | "appendFile">;

interface SessionHeader {
	type: "session";
	version: 3;
	id: string;
	timestamp: string;
	cwd: string;
	parentSession?: string;
	metadata?: Record<string, unknown>;
}

/**
 * 使用单个条目的标签数据更新内存中的标签查找映射。
 * 仅作用于类型为 `"label"` 的条目。如果标签文本非空，则将目标条目 ID 映射到该标签；
 * 如果标签为空或 `undefined`，则移除该目标的任何现有映射。
 * @param labelsById - 从目标条目 ID 到其当前标签的可变映射。
 * @param entry - 会话树条目，可能是也可能不是标签条目。
 */
function updateLabelCache(labelsById: Map<string, string>, entry: SessionTreeEntry): void {
	if (entry.type !== "label") return;
	const label = entry.label?.trim();
	if (label) {
		labelsById.set(entry.targetId, label);
	} else {
		labelsById.delete(entry.targetId);
	}
}

/**
 * 从会话树条目数组构建完整的标签查找映射。
 * 遍历所有条目并累积每个目标 ID 的最新标签。后出现的标签条目会覆盖先出现的。
 * @param entries - 要扫描标签条目的会话树条目数组。
 * @returns 从条目 ID 到其当前标签字符串的映射。
 */
function buildLabelsById(entries: SessionTreeEntry[]): Map<string, string> {
	const labelsById = new Map<string, string>();
	for (const entry of entries) {
		updateLabelCache(labelsById, entry);
	}
	return labelsById;
}

/**
 * 生成一个保证不与现有 ID 冲突的唯一条目 ID。
 * 最多尝试 100 次生成 UUIDv7 的 8 字符后缀，确保其在提供的查找表中不存在。
 * @param byId - 具有 `has(id)` 方法用于检查 ID 冲突的对象。
 * @returns 新会话条目的唯一字符串 ID。
 */
function generateEntryId(byId: { has(id: string): boolean }): string {
	for (let i = 0; i < 100; i++) {
		// The uuidv7 prefix is timestamp-derived and nearly constant between calls,
		// so short ids must come from the random tail.
		const id = uuidv7().slice(-8);
		if (!byId.has(id)) return id;
	}
	return uuidv7();
}

/**
 * 构造一个代码为 `"invalid_session"` 的 {@link SessionError}。
 * 当 JSONL 会话文件的头部或整体结构格式错误或缺少必需字段时使用。
 * @param filePath - 有问题的会话文件路径。
 * @param message - 验证失败的简短描述。
 * @param cause - 触发此失败的可选底层错误。
 * @returns 可抛出的 {@link SessionError}。
 */
function invalidSession(filePath: string, message: string, cause?: Error): SessionError {
	return new SessionError("invalid_session", `Invalid JSONL session file ${filePath}: ${message}`, cause);
}

/**
 * 构造一个代码为 `"invalid_entry"` 的 {@link SessionError}。
 * 当 JSONL 会话文件中的特定行未通过结构验证时使用。
 * @param filePath - 包含无效条目的会话文件路径。
 * @param lineNumber - 发现无效条目的行号（从 1 开始）。
 * @param message - 验证失败的简短描述。
 * @param cause - 触发此失败的可选底层错误。
 * @returns 可抛出的 {@link SessionError}。
 */
function invalidEntry(filePath: string, lineNumber: number, message: string, cause?: Error): SessionError {
	return new SessionError(
		"invalid_entry",
		`Invalid JSONL session file ${filePath}: line ${lineNumber} ${message}`,
		cause,
	);
}

/**
 * 解析并验证 JSONL 会话文件的第一行作为会话头部。有效的头部必须包含 type、version、id、timestamp、cwd 字段。
 * @param line - JSONL 文件的原始第一行（未修剪）。
 * @param filePath - 正在解析的文件路径（用于错误消息）。
 * @returns 已验证的 {@link SessionHeader} 对象。
 * @throws {SessionError} 如果验证失败，代码为 `"invalid_session"`。
 */
function parseHeaderLine(line: string, filePath: string): SessionHeader {
	let parsed: unknown;
	try {
		parsed = JSON.parse(line);
	} catch (error) {
		throw invalidSession(filePath, "first line is not a valid session header", toError(error));
	}
	if (typeof parsed !== "object" || parsed === null) {
		throw invalidSession(filePath, "first line is not a valid session header");
	}
	const header = parsed as Partial<SessionHeader>;
	if (header.type !== "session") throw invalidSession(filePath, "first line is not a valid session header");
	if (header.version !== 3) throw invalidSession(filePath, "unsupported session version");
	if (typeof header.id !== "string" || !header.id) throw invalidSession(filePath, "session header is missing id");
	if (typeof header.timestamp !== "string" || !header.timestamp) {
		throw invalidSession(filePath, "session header is missing timestamp");
	}
	if (typeof header.cwd !== "string" || !header.cwd) throw invalidSession(filePath, "session header is missing cwd");
	if (header.parentSession !== undefined && typeof header.parentSession !== "string") {
		throw invalidSession(filePath, "session header parentSession must be a string");
	}
	if (
		header.metadata !== undefined &&
		(typeof header.metadata !== "object" || header.metadata === null || Array.isArray(header.metadata))
	) {
		throw invalidSession(filePath, "session header metadata must be an object");
	}
	return {
		type: "session",
		version: 3,
		id: header.id,
		timestamp: header.timestamp,
		cwd: header.cwd,
		parentSession: header.parentSession,
		metadata: header.metadata,
	};
}

/**
 * 解析并验证单行（头部之后的行）作为会话树条目。
 * 每个条目行必须是一个 JSON 对象，至少包含 type、id、parentId、timestamp 字段。
 * @param line - 要解析的原始 JSON 行。
 * @param filePath - 文件路径（用于错误消息）。
 * @param lineNumber - 文件内的行号（从 1 开始）。
 * @returns 解析后的条目，强制转换为 {@link SessionTreeEntry}。
 * @throws {SessionError} 如果验证失败，代码为 `"invalid_entry"`。
 */
function parseEntryLine(line: string, filePath: string, lineNumber: number): SessionTreeEntry {
	let parsed: unknown;
	try {
		parsed = JSON.parse(line);
	} catch (error) {
		throw invalidEntry(filePath, lineNumber, "is not valid JSON", toError(error));
	}
	if (typeof parsed !== "object" || parsed === null) {
		throw invalidEntry(filePath, lineNumber, "is not a valid session entry");
	}
	const entry = parsed as {
		type?: unknown;
		id?: unknown;
		parentId?: unknown;
		timestamp?: unknown;
		targetId?: unknown;
	};
	if (typeof entry.type !== "string") throw invalidEntry(filePath, lineNumber, "is missing entry type");
	if (typeof entry.id !== "string" || !entry.id) throw invalidEntry(filePath, lineNumber, "is missing entry id");
	if (entry.parentId !== null && typeof entry.parentId !== "string") {
		throw invalidEntry(filePath, lineNumber, "has invalid parentId");
	}
	if (typeof entry.timestamp !== "string" || !entry.timestamp) {
		throw invalidEntry(filePath, lineNumber, "is missing timestamp");
	}
	if (entry.type === "leaf" && entry.targetId !== null && typeof entry.targetId !== "string") {
		throw invalidEntry(filePath, lineNumber, "has invalid targetId");
	}
	return entry as SessionTreeEntry;
}

/**
 * 确定处理给定条目后的有效叶节点 ID。
 * 对于叶节点条目（`type === "leaf"`），叶节点 ID 是条目的 `targetId`；
 * 对于所有其他条目类型，叶节点 ID 是条目自身的 ID。
 * @param entry - 刚被追加或处理的条目。
 * @returns 新的叶节点 ID，如果树指向根节点则为 `null`。
 */
function leafIdAfterEntry(entry: SessionTreeEntry): string | null {
	return entry.type === "leaf" ? entry.targetId : entry.id;
}

/**
 * 将解析后的会话头部及其文件路径转换为 {@link JsonlSessionMetadata}。
 * 提取会话 ID、创建时间戳、工作目录、可选的父会话路径以及 JSONL 文件路径本身。
 * @param header - 从 JSONL 文件解析出的已验证 {@link SessionHeader}。
 * @param path - 会话 JSONL 文件的绝对路径。
 * @returns {@link JsonlSessionMetadata} 对象。
 */
function headerToSessionMetadata(header: SessionHeader, path: string): JsonlSessionMetadata {
	return {
		id: header.id,
		createdAt: header.timestamp,
		cwd: header.cwd,
		path,
		parentSessionPath: header.parentSession,
		metadata: header.metadata,
	};
}

export async function loadJsonlSessionMetadata(
	fs: JsonlSessionStorageFileSystem,
	filePath: string,
): Promise<JsonlSessionMetadata> {
	const lines = getFileSystemResultOrThrow(
		await fs.readTextLines(filePath, { maxLines: 1 }),
		`Failed to read session header ${filePath}`,
	);
	const line = lines[0];
	if (line?.trim()) return headerToSessionMetadata(parseHeaderLine(line, filePath), filePath);
	throw invalidSession(filePath, "missing session header");
}

/**
 * 从磁盘加载并解析整个 JSONL 会话文件到其内存表示。
 * 读取完整文件内容，解析头部和所有条目行，跟踪当前叶节点 ID。
 * @param fs - 用于读取文件的文件系统抽象。
 * @param filePath - JSONL 会话文件的绝对路径。
 * @returns 包含解析后的头部、所有条目的有序数组以及当前叶节点 ID 的对象。
 * @throws {SessionError} 如果文件缺失、为空或包含无效行。
 */
async function loadJsonlStorage(
	fs: JsonlSessionStorageFileSystem,
	filePath: string,
): Promise<{
	header: SessionHeader;
	entries: SessionTreeEntry[];
	leafId: string | null;
}> {
	const content = getFileSystemResultOrThrow(await fs.readTextFile(filePath), `Failed to read session ${filePath}`);
	const lines = content.split("\n").filter((line) => line.trim());
	if (lines.length === 0) {
		throw invalidSession(filePath, "missing session header");
	}

	const header = parseHeaderLine(lines[0]!, filePath);
	const entries: SessionTreeEntry[] = [];
	let leafId: string | null = null;
	for (let i = 1; i < lines.length; i++) {
		const entry = parseEntryLine(lines[i]!, filePath, i + 1);
		entries.push(entry);
		leafId = leafIdAfterEntry(entry);
	}
	return { header, entries, leafId };
}

export class JsonlSessionStorage implements SessionStorage<JsonlSessionMetadata> {
	private readonly fs: JsonlSessionStorageFileSystem;
	private readonly filePath: string;
	private readonly metadata: JsonlSessionMetadata;
	private entries: SessionTreeEntry[];
	private byId: Map<string, SessionTreeEntry>;
	private labelsById: Map<string, string>;
	private currentLeafId: string | null;

	private constructor(
		fs: JsonlSessionStorageFileSystem,
		filePath: string,
		header: SessionHeader,
		entries: SessionTreeEntry[],
		leafId: string | null,
	) {
		this.fs = fs;
		this.filePath = filePath;
		this.metadata = headerToSessionMetadata(header, this.filePath);
		this.entries = entries;
		this.byId = new Map(entries.map((entry) => [entry.id, entry]));
		this.labelsById = buildLabelsById(entries);
		this.currentLeafId = leafId;
	}

	static async open(fs: JsonlSessionStorageFileSystem, filePath: string): Promise<JsonlSessionStorage> {
		const loaded = await loadJsonlStorage(fs, filePath);
		return new JsonlSessionStorage(fs, filePath, loaded.header, loaded.entries, loaded.leafId);
	}

	static async create(
		fs: JsonlSessionStorageFileSystem,
		filePath: string,
		options: {
			cwd: string;
			sessionId: string;
			parentSessionPath?: string;
			metadata?: Record<string, unknown>;
		},
	): Promise<JsonlSessionStorage> {
		const header: SessionHeader = {
			type: "session",
			version: 3,
			id: options.sessionId,
			timestamp: new Date().toISOString(),
			cwd: options.cwd,
			parentSession: options.parentSessionPath,
			metadata: options.metadata,
		};
		getFileSystemResultOrThrow(
			await fs.writeFile(filePath, `${JSON.stringify(header)}\n`),
			`Failed to create session ${filePath}`,
		);
		return new JsonlSessionStorage(fs, filePath, header, [], null);
	}

	async getMetadata(): Promise<JsonlSessionMetadata> {
		return this.metadata;
	}

	async getLeafId(): Promise<string | null> {
		if (this.currentLeafId !== null && !this.byId.has(this.currentLeafId)) {
			throw new SessionError("invalid_session", `Entry ${this.currentLeafId} not found`);
		}
		return this.currentLeafId;
	}

	async setLeafId(leafId: string | null): Promise<LeafEntry> {
		if (leafId !== null && !this.byId.has(leafId)) {
			throw new SessionError("not_found", `Entry ${leafId} not found`);
		}
		const entry: LeafEntry = {
			type: "leaf",
			id: generateEntryId(this.byId),
			parentId: this.currentLeafId,
			timestamp: new Date().toISOString(),
			targetId: leafId,
		};
		getFileSystemResultOrThrow(
			await this.fs.appendFile(this.filePath, `${JSON.stringify(entry)}\n`),
			`Failed to append session leaf ${entry.id}`,
		);
		this.entries.push(entry);
		this.byId.set(entry.id, entry);
		this.currentLeafId = leafId;
		return entry;
	}

	async createEntryId(): Promise<string> {
		return generateEntryId(this.byId);
	}

	async appendEntry(entry: SessionTreeEntry): Promise<void> {
		getFileSystemResultOrThrow(
			await this.fs.appendFile(this.filePath, `${JSON.stringify(entry)}\n`),
			`Failed to append session entry ${entry.id}`,
		);
		this.entries.push(entry);
		this.byId.set(entry.id, entry);
		updateLabelCache(this.labelsById, entry);
		this.currentLeafId = leafIdAfterEntry(entry);
	}

	async getEntry(id: string): Promise<SessionTreeEntry | undefined> {
		return this.byId.get(id);
	}

	async findEntries<TType extends SessionTreeEntry["type"]>(
		type: TType,
	): Promise<Array<Extract<SessionTreeEntry, { type: TType }>>> {
		return this.entries.filter((entry): entry is Extract<SessionTreeEntry, { type: TType }> => entry.type === type);
	}

	async getLabel(id: string): Promise<string | undefined> {
		return this.labelsById.get(id);
	}

	/** 返回最近的会话名称（如果有记录）。 */
	async getSessionName(): Promise<string | undefined> {
		const entries = await this.findEntries("session_info");
		return entries[entries.length - 1]?.name?.trim() || undefined;
	}

	/** 计算会话统计信息（消息数量、缓存/非缓存 token 数、总费用）。 */
	async getSessionStats() {
		let messageCount = 0;
		let cachedTokens = 0;
		let uncachedTokens = 0;
		let totalTokens = 0;
		let costTotal = 0;
		for (const entry of this.entries) {
			if (entry.type === "message") {
				messageCount += 1;
			}
			const usage =
				entry.type === "message"
					? entry.message.role === "assistant"
						? entry.message.usage
						: undefined
					: entry.type === "compaction" || entry.type === "branch_summary"
						? entry.usage
						: undefined;
			if (
				!usage ||
				typeof usage.input !== "number" ||
				typeof usage.output !== "number" ||
				typeof usage.cacheRead !== "number" ||
				typeof usage.cacheWrite !== "number" ||
				typeof usage.cost?.total !== "number"
			) {
				continue;
			}
			cachedTokens += usage.cacheRead;
			uncachedTokens += usage.input + usage.cacheWrite;
			totalTokens += usage.input + usage.output + usage.cacheRead + usage.cacheWrite;
			costTotal += usage.cost.total;
		}
		return {
			messageCount,
			cachedTokens,
			uncachedTokens,
			totalTokens,
			costTotal,
		};
	}

	/** 从叶节点到根或到压缩边界的路径。遇到带有 retainedTail 的压缩条目时停止。 */
	async getPathToRootOrCompaction(leafId: string | null): Promise<SessionTreeEntry[]> {
		if (leafId === null) return [];
		const path: SessionTreeEntry[] = [];
		let stopAtEntryId: string | null = null;
		let current = this.byId.get(leafId);
		if (!current) throw new SessionError("not_found", `Entry ${leafId} not found`);
		while (current) {
			path.unshift(current);
			if (stopAtEntryId !== null && current.id === stopAtEntryId) break;
			if (current.type === "compaction") {
				if (current.retainedTail) break;
				stopAtEntryId = current.firstKeptEntryId ?? null;
			}
			if (!current.parentId) break;
			const parent = this.byId.get(current.parentId);
			if (!parent) throw new SessionError("invalid_session", `Entry ${current.parentId} not found`);
			current = parent;
		}
		return path;
	}

	/** 支持游标分页地返回条目的子集。 */
	async getEntries(options?: SessionEntryCursorOptions): Promise<SessionTreeEntry[]> {
		const start = options?.afterEntrySeq ?? 0;
		const end = options?.limit === undefined ? undefined : start + options.limit;
		return this.entries.slice(start, end);
	}
}
